clients/spark/core/src/main/scala/io/treeverse/clients/ApiClient.scala

```diff
@@ -198,28 +219,48 @@ class ApiClient private (conf: APIConfigurations) {
    *  translate that URL to use an appropriate Hadoop FileSystem.
    */
   def getMetaRangeURL(repoName: String, commitID: String): String = {
-    val commit = commitsApi.getCommit(repoName, commitID)
+    val commit = getCommit(repoName, commitID)
     getMetaRangeURL(repoName, commit)
   }
 
   /** Query lakeFS for a URL to the range of rangeID of repoName and
    *  translate that URL to use an appropriate Hadoop FileSystem.
    */
   def getRangeURL(repoName: String, rangeID: String): String = {
-    val range = metadataApi.getRange(repoName, rangeID)
+    val getRange = new dev.failsafe.function.CheckedSupplier[StorageURI]() {
+      def get(): StorageURI = metadataApi.getRange(repoName, rangeID)
+    }
+    val range = retryWrapper.wrapWithRetry(getRange)
     val location = range.getLocation
     URI
       .create(getStorageNamespace(repoName, StorageClientType.HadoopFS) + "/" + location)
       .normalize()
       .toString
   }
 
-  def getBranchHEADCommit(repoName: String, branch: String): String =
-    branchesApi.getBranch(repoName, branch).getCommitId
+  def getBranchHEADCommit(repoName: String, branch: String): String = {
+    val getBranch = new dev.failsafe.function.CheckedSupplier[Ref]() {
+      def get(): Ref = branchesApi.getBranch(repoName, branch)
+    }
+    val response = retryWrapper.wrapWithRetry(getBranch)
+    response.getCommitId
+  }
 
   // Instances of case classes are compared by structure and not by reference https://docs.scala-lang.org/tour/case-classes.html.
   case class StorageNamespaceCacheKey(
       repoName: String,
       storageClientType: StorageClientType
   )
 }
+
+class RequestRetryWrapper(val numRetries: Int = 5) {
+    private val retryPolicy : Policy[Any] = RetryPolicy.builder()
+      .withDelay(Duration.ofMillis(500))
```

Comment:
* I think that exponential backoff makes sense here rather than this linear/constant backoff, at least for the massive calls like `getRange`.  Otherwise I worry that a Spark cluster with a lot of processing power can still overload a weak lakeFS cluster.
* Does failsafe automatically add jitter to its calls?  If not then we should ask it to add jitter.  Otherwise all retries from a cluster can happen almost exactly at once!  (If the package offers other guidance, then of course we should go with what the experts say...).
* It offers a total of 2s backoff, which is not a lot for a Spark job that runs for very many minutes.  I'd be perfectly happy with waiting up to 20 seconds (say) before deciding that this job has failed).
* OTOH, this makes the read timeout configuration _really hard to use_!  If I configure a read timeout of 10s, then it waits in total for 5*10s + 4*0.5s = 52s.  That will not be very clear to users.  Two possible fixes (others are possible!) might be to expose all parameters, or have 2 parameters `read_timeout` and `total_read_timeout` and somehow tell failsafe to handle retries that take up to `total_read_timeout`.